keps/sig-api-machinery/20190329-less-object-serializations.md

```diff
@@ -0,0 +1,258 @@
+by kube-proxy running on every one) kube-apiserver independently serializes
+(which also requires deep-copy) every single object being send via this watch.
+
+While this problem is extremely visible for watch, the situation looks the same
+for regular GET/LIST operations - reading the same object N times will result
+in serializing that N times independently.
+
+This proposal presents a solution for that problem.
+
+[#75294]: https://github.com/kubernetes/kubernetes/issues/75294
+
+### Goals
+
+- Reduce load on kube-apiserver, by not serializing the same object multiple
+times wherever it is possible.
+
+### Non-Goals
+
+- Change overall architecture of the system, by changing what data is being
+read/watched by different components.
+
+## Proposal
+
+This proposal does not introduce any user-visible changes - the proposed changes
+are purely implementational detail of kube-apiserver.
+
+We propose introducing an internal type implementing `runtime.Object` interface,
+that is encapsulating the original object and is additionally able to accumulate
+its serialized versions. It will look like this:
+```
+// TODO: Better name is welcome :)
+type CachingObject struct {
+	// Object is the object (potentially in the internal version)
+	// for which serializations should be cached for future reuse.
+	Object Object
+	// Versioned is a list of cached versioned serializations.
+	// Note that a list is needed, because a given object can be serialized
+	// to different external versions.
+	Versioned []*CachingVersionedObject
+}
+
+// TODO: Better name is welcome :)
+type CachingVersionedObject struct {
+	// Object in an versioned object.
+	Object Object
+	// Serializations is a list of cached serializations of Object.
+	// Note that a list is needed, because a given versioned object can be
+	// serialized to different media types.
+	//
+	// TODO: Maybe we should be more explicit and just store one for json
+	// and one for protobuf?
+	Serializations []*serializedObject
+}
+
+// TODO: Better name is welcome :)
+type serializedObject struct {
+	once sync.Once
+	// mediaType to which the object is serialized.
+	mediaType string
+	// raw is serialized object.
+	raw []byte
+	// err is error from serialization.
+	err error
+}
+```
+Note that both `CachingObject` and `CachingVersionedObject` should implement
+`runtime.Object` interface.
+
+For those two objects, we will implement `Marshal()`, `MarshallTo()` (for
+protobuf) and `MarshalJSON()` (for json) methods, that:
+- first check if such serialization has already been done, and if so just
+  returning the cached result
+- otherwise performing the serialization, caching the result and returning
+  it to the called
+
+With that, we will need to modify `versioning.codec` so that when encoding
+`CachingObject`, it will perform necessary conversion if needed to the
+`CachingVersionedObject`.
+
+Now the question is what is responsible for creating `CachingObject`.
```

Comment:
I think I prefer the smart object idea over trying to externally manage a different cache.

Reply:
smart object meaning the cached object? Or the proto-aware object I described?

Reply:
I added "smart object" to the "alternatives" section.